tools/ci/src/main.rs

```diff
@@ -44,125 +61,300 @@ fn main() {
+                command: cmd!(sh, "cargo check {args...}"),
+                failure_message: "Please fix compiler errors for examples in output above.",
+                subdir: None,
+                env_vars: vec![],
+            }],
+        );
     }
 
-    if what_to_run.contains(Check::COMPILE_CHECK) {
+    if checks.contains(Check::COMPILE_CHECK) {
         // Build bevy and check that it compiles
-        cmd!(sh, "cargo check --workspace")
-            .run()
-            .expect("Please fix compiler errors in output above.");
+        let mut args = vec!["--workspace"];
+        if flags.contains(Flag::KEEP_GOING) {
+            args.push("--keep-going");
+        }
+
+        test_suite.insert(
+            Check::COMPILE_CHECK,
+            vec![CITest {
+                command: cmd!(sh, "cargo check {args...}"),
+                failure_message: "Please fix compiler errors in output above.",
+                subdir: None,
+                env_vars: vec![],
+            }],
+        );
     }
 
-    if what_to_run.contains(Check::CFG_CHECK) {
+    if checks.contains(Check::CFG_CHECK) {
         // Check cfg and imports
-        std::env::set_var("RUSTFLAGS", "-D warnings");
-        cmd!(sh, "cargo +nightly check -Zcheck-cfg --workspace")
-            .run()
-            .expect("Please fix failing cfg checks in output above.");
+        let mut args = vec!["-Zcheck-cfg", "--workspace"];
+        if flags.contains(Flag::KEEP_GOING) {
+            args.push("--keep-going");
+        }
+
+        test_suite.insert(
+            Check::CFG_CHECK,
+            vec![CITest {
+                command: cmd!(sh, "cargo +nightly check {args...}"),
+                failure_message: "Please fix failing cfg checks in output above.",
+                subdir: None,
+                env_vars: vec![("RUSTFLAGS", "-D warnings")],
+            }],
+        );
+    }
+
+    // Actually run the tests:
+
+    let mut failed_checks: Check = Check::empty();
+    let mut failure_message: String = String::new();
+
+    // In KEEP_GOING-mode, we save all errors until the end; otherwise, we just
+    // panic with the given message for test failure.
+    fn fail(
+        current_check: Check,
+        failure_message: &'static str,
+        failed_checks: &mut Check,
+        existing_fail_message: &mut String,
+        flags: &Flag,
+    ) {
+        if flags.contains(Flag::KEEP_GOING) {
+            failed_checks.insert(current_check);
+            if !existing_fail_message.is_empty() {
+                existing_fail_message.push('\n');
+            }
+            existing_fail_message.push_str(failure_message);
+        } else {
+            panic!("{failure_message}");
+        }
+    }
+
+    for (check, battery) in test_suite.into_iter() {
+        for ci_test in battery {
+            // Ensure that necessary environment variables are set
+            for (k, v) in ci_test.env_vars {
+                std::env::set_var(k, v);
+            }
```

Comment:
This will cause environmental variables to "leak" into other test suites. If for instance `PATH=../path/to/exe` was set for one test suite, it may mess up other test suites that also use the path.

Reply:
This is a good point! The original implementation did this too, but I should probably do better if I'm going to the effort of refactoring it. I'll throw in a little drop-hook abstraction to take care of this.

Reply:
You may be interested in [`Command::env`], which only sets the environmental variable for the child process. That way you don't have to worry about accidentally clearing one for the CI program.

[`Command::env`]: https://doc.rust-lang.org/stable/std/process/struct.Command.html#method.env

Reply:
Looks like xshell even lifts that command to `Cmd::env`, which should make using it a breeze! I'll switch my solution to use that later today (probably).